src/KubernetesClient/Authentication/OidcTokenProvider.cs

```diff
@@ -14,23 +16,33 @@ public class OidcTokenProvider : ITokenProvider
         private string _idToken;
         private string _refreshToken;
         private string _accessToken;
-        private DateTime _expiry;
+        private DateTimeOffset _expiry;
 
         public OidcTokenProvider(string clientId, string clientSecret, string idpIssuerUrl, string idToken, string refreshToken)
         {
             _idToken = idToken;
             _refreshToken = refreshToken;
             _oidcClient = getClient(clientId, clientSecret, idpIssuerUrl);
+            _expiry = getExpiryFromToken();
         }
 
         public async Task<AuthenticationHeaderValue> GetAuthenticationHeaderAsync(CancellationToken cancellationToken)
         {
-            if (_accessToken == null || DateTime.UtcNow.AddSeconds(30) > _expiry)
+            if ((_accessToken == null && _idToken == null) || DateTimeOffset.UtcNow.AddSeconds(30) > _expiry)
```

Comment:
any case that access token is null and idtoken is not null?
and i think we need to double check if idtoken has access to resources

Reply:
>any case that access token is null and idtoken is not null?

yes, that is the case where i was having an issue. my access token was null but i had an unexpired id token and the client would try to refresh when i didnt need to refresh. the unneeded refresh would fail due to our policies and no successful request could be made

>and i think we need to double check if idtoken has access to resources

im not sure i understand. how would we go about checking that? why would that check need to be added now when it wasnt needed before? is the javascript client doing this kind of checking?

Reply:
not 100% sure if the `id token` can be used as `bearer`
checked <https://openid.net/specs/openid-connect-core-1_0.html> and did not see if the `idtoken` can be used as the `bearer`

however, I also checked <https://github.com/kubernetes-client/javascript/blob/681af0941f306a1f6ac6a4d7622a5901c5ebb0e3/src/oidc_auth.ts#L68>
seems js is using idtoken only

my concern is if we are following the spec

Reply:
a lot of these concepts are new to me but from what i can tell the javascript implementation of only using id token is correct. i started down this path initially but then was worried about breaking things for @arbielsk who implemented using the access token. im fairly confident now that we can remove the access token
>To identify the user, the authenticator uses the `id_token` (not the `access_token`) from the OAuth2 token response as a bearer token.

https://kubernetes.io/docs/reference/access-authn-authz/authentication/#openid-connect-tokens